Write post about 7.1.0.beta1 release (#180)
rails/website: _posts/2023-09-11-Rails-7-1-0-beta-1-has-been-released.markdown

    password_salt&.last(10)
end

user = User.first

token = user.generate_token_for(:password_reset)
User.find_by_token_for(:password_reset, token) # => user
# 16 minutes later...
User.find_by_token_for(:password_reset, token) # => nil
```

Finally, [`has_secure_password`](https://edgeapi.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password)
got the ability to automatically verify the current password when updating the password. This is useful to check
if the user who is trying to update the password, knows the current password.

```ruby
# Schema: User(name:string, password_digest:string)
class User < ActiveRecord::Base
  has_secure_password
end

user = User.new(name: "rafael", password: "railsrocks", password_confirmation: "railsrocks")
user.save                                                                      # => true
user.update(password: "pwn3d", password_challenge: "")                         # => false, challenge doesn't authenticate
user.update(password: "railsGetsEvenBetter", password_challenge: "railsrocks") # => true
```

## A more robust framework

> Rails promises to be a framework that can get you from Hello World to IPO, and on each release we are improving

Author: @dhh I tried to see if we wrote a blog post about the "Hello World to IPO" but I can only find your tweets. Do you have any post in your blog about this that we can link here?

---
layout: post
title: 'Rails 7.1.0 Beta 1: Dockerfiles, Attribute Normalization, Single Use Tokens and More'
categories: releases
author: rafaelfranca
published: true
date: 2023-09-11 21:00:00 -04:00
---

> RailsWorld is going to be soon, and to prepare for the celebration of the 20th anniversary of Rails, we are releasing the first beta of Rails 7.1.0!

Reviewer: ```suggestion
RailsWorld is approaching., and to prepare for the celebration of the 20th anniversary of Rails, we are releasing the first beta of Rails 7.1.0!
```

There are [over five thousand commits](https://github.com/rails/rails/compare/v7.0.8...main) since Rails 7.0, so this release is packed with new features and improvements.

Please help us test all this new stuff so we can ensure a solid final release of Rails 7.1.

> ## Generate Dockerfile for new applications

Reviewer: ```suggestion
## Generate Dockerfiles for new applications
```

Rails will now generate a Docker-related files when you run `rails new`.

> These files serve as a foundational setup for deploying your Rails application in a

Reviewer: ```suggestion
These files serve as foundational setup for deploying your Rails application in a
```

production environment using Docker.

## Building blocks for authentication

To complement with `has_secure_password` Rails 7.1 brings a few new building blocks to help developers to build their
own authentication system.

First, [`normalizes`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Normalization/ClassMethods.html#method-i-normalizes)
declares an attribute normalization. This is useful to normalize attributes before saving them to the database.

```ruby
class User < ApplicationRecord
  normalizes :email, with: -> email { email.strip.downcase }
end

> user = User.create(email: " CRUISE-CONTROL@EXAMPLE.COM\n")

Reviewer: ```suggestion
user = User.create(email: "CRUISE-CONTROL@EXAMPLE.COM\n")
```

Author: Space if by design. We are showing the strip call

Reviewer: Aaaaaa I missed that! 👍🏾

user.update(password: "pwn3d", password_challenge: "")                         # => false, challenge doesn't authenticate
user.update(password: "railsGetsEvenBetter", password_challenge: "railsrocks") # => true
```

## A more robust framework

Rails promises to be a framework that can get you from Hello World to IPO, and on each release we are improving
how robust Rails is.

For 7.1 we are bringing a collection of changes that have the goal of making easier for your application to scale.

The [`perform_all_later` method in Active Job](https://edgeapi.rubyonrails.org/classes/ActiveJob.html#method-c-perform_all_later),
was designed to streamline the process of enqueuing multiple jobs simultaneously. This powerful
addition allows you to efficiently enqueue jobs without triggering callbacks. This is
particularly useful when you need to enqueue a batch of jobs at once, reducing the overhead
of multiple round-trips to the queue datastore.

Support to Composite Primary keys was added to Active Record, [allowing applications to improve the performance of some
queries](https://shopify.engineering/how-to-introduce-composite-primary-keys-in-rails), specially those applications

> that use multi-tenant architecture, by using the tenant id as part of the primary key.

Reviewer: ```suggestion
that use multi-tenant architecture, by using the tenant ID as part of the primary key.
```

or

```suggestion
that use multi-tenant architecture, by using the tenant `id` as part of the primary key.
```

Reviewer 2: Agree with this suggestion! I'd say the second option could be `tenant_id` if we want to imply a particular column name or "tenant ID" if we want to imply "tenant identifier"

A new database adapter was added, [Trilogy](https://github.blog/2022-08-25-introducing-trilogy-a-new-database-adapter-for-ruby-on-rails/),
a MySQL-compatible database client. This adapter is mostly a drop-in replacement for the MySQL2 adapter, but streamlines
the installation process by not requiring the mysql client library to

> be present in the operation system, and improve the performance of some queries.

Reviewer: ```suggestion
be present in the operating system, and improve the performance of some queries.
```
